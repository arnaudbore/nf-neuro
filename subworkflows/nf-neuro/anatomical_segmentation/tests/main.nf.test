nextflow_workflow {

    name "Test Subworkflow ANATOMICAL_SEGMENTATION"
    script "../main.nf"
    workflow "ANATOMICAL_SEGMENTATION"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfneuro"
    tag "subworkflows/anatomical_segmentation"

    tag "segmentation"
    tag "segmentation/fastseg"
    tag "segmentation/freesurferseg"
    tag "segmentation/synthseg"

    tag "load_test_data"

    tag "stub"
    options "-stub-run"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "T1w.zip", "freesurfer_nifti.zip", "freesurfer.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("anatomical_segmentation - fslfast") {
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        freesurfer_nifti: it.simpleName == "freesurfer_nifti"
                        freesurfer: it.simpleName == "freesurfer"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz", checkIfExists: true)
                    ]
                }
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value  }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    [ subject[0] ] + subject[1..-1].collect{ entry -> entry ? file(entry).name : "" }
                            } ]
                        }
                ).match() },
                { assert workflow.out
                    .findAll{ channel -> !channel.key.isInteger() }
                    .every{ channel -> ["wm_mask",
                                        "gm_mask",
                                        "csf_mask",
                                        "wm_map",
                                        "gm_map",
                                        "csf_map",
                                        "versions"].contains(channel.key)
                        ? channel.value.every{ subject -> subject instanceof ArrayList
                                ? subject.every()
                                : subject }
                        : channel.value.size() == 0 } }
            )
        }
    }

    test("anatomical_segmentation - freesurferseg") {
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        freesurfer_nifti: it.simpleName == "freesurfer_nifti"
                        freesurfer: it.simpleName == "freesurfer"
                    }
                input[0] = Channel.empty()
                input[1] = ch_split_test_data.freesurfer_nifti.map{
                    test_data_directory -> [
                        [ id:'test' ], // meta map
                        file("\${test_data_directory}/aparc+aseg.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/wmparc.nii.gz", checkIfExists: true)
                    ]
                }
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value  }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    [ subject[0] ] + subject[1..-1].collect{ entry -> entry ? file(entry).name : "" }
                            } ]
                        }
                ).match() },
                { assert workflow.out
                    .findAll{ !it.key.isInteger() }
                    .every{ channel -> ["wm_mask",
                                        "gm_mask",
                                        "csf_mask",
                                        "versions"].contains(channel.key)
                        ? channel.value.every{ subject -> subject instanceof ArrayList
                                ? subject.every()
                                : subject }
                        : channel.value.size() == 0 } }
            )
        }
    }

    test("anatomical_segmentation - both") {
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        freesurfer_nifti: it.simpleName == "freesurfer_nifti"
                        freesurfer: it.simpleName == "freesurfer"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz", checkIfExists: true)
                    ]
                }
                input[1] = ch_split_test_data.freesurfer_nifti.map{
                    test_data_directory -> [
                        [ id:'test' ], // meta map
                        file("\${test_data_directory}/aparc+aseg.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/wmparc.nii.gz", checkIfExists: true)
                    ]
                }
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value  }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    [ subject[0] ] + subject[1..-1].collect{ entry -> entry ? file(entry).name : "" }
                            } ]
                        }
                ).match() },
                { assert workflow.out
                    .findAll{ !it.key.isInteger() }
                    .every{ channel -> ["wm_mask",
                                        "gm_mask",
                                        "csf_mask",
                                        "wm_map",
                                        "gm_map",
                                        "csf_map",
                                        "versions"].contains(channel.key)
                        ? channel.value.every{ subject -> subject instanceof ArrayList
                                ? subject.every()
                                : subject }
                        : channel.value.size() == 0 } }
            )
        }
    }

    test("anatomical_segmentation - synthseg") {
        config "./synthseg.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        t1w: it.simpleName == "T1w"
                        freesurfer_nifti: it.simpleName == "freesurfer_nifti"
                        freesurfer: it.simpleName == "freesurfer"
                    }
                input[0] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz", checkIfExists: true)
                    ]
                }
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                input[3] = ch_split_test_data.freesurfer.map{
                    test_data_directory -> [
                        file("\${test_data_directory}/license.txt", checkIfExists: true)
                    ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value  }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    [ subject[0] ] + subject[1..-1].collect{ entry -> entry ? file(entry).name : "" }
                            } ]
                        }
                ).match() },
                { assert workflow.out
                    .findAll{ !it.key.isInteger() }
                    .every{ channel -> ["wm_mask",
                                        "gm_mask",
                                        "csf_mask",
                                        "wm_map",
                                        "gm_map",
                                        "csf_map",
                                        "aparc_aseg",
                                        "qc_score",
                                        "resample",
                                        "seg",
                                        "volume",
                                        "versions"].contains(channel.key)
                        ? channel.value.every{ subject -> subject instanceof ArrayList
                                ? subject.every()
                                : subject }
                        : channel.value.size() == 0 } }
            )
        }
    }
}
