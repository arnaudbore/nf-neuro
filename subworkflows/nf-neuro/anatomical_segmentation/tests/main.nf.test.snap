{
    "anatomical_segmentation - freesurferseg": {
        "content": [
            {
                "csf_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_csf.nii.gz"
                    ]
                ],
                "gm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_gm.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,7f184d3df98a22ec61c2e52230199180"
                ],
                "wm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_wm.nii.gz"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-30T16:36:29.593552"
    },
    "anatomical_segmentation - synthseg": {
        "content": [
            {
                "aparc_aseg": [
                    [
                        {
                            "id": "test"
                        },
                        "test__aparc_aseg.nii.gz"
                    ]
                ],
                "csf_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_csf.nii.gz"
                    ]
                ],
                "csf_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_csf.nii.gz"
                    ]
                ],
                "gm_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_gm.nii.gz"
                    ]
                ],
                "gm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_gm.nii.gz"
                    ]
                ],
                "qc_score": [
                    [
                        {
                            "id": "test"
                        },
                        "test__qc_score.csv"
                    ]
                ],
                "resample": [
                    [
                        {
                            "id": "test"
                        },
                        "test__resampled_image.nii.gz"
                    ]
                ],
                "seg": [
                    [
                        {
                            "id": "test"
                        },
                        "test__seg.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,95e31db072c7bb2505128bd08beaecf4"
                ],
                "volume": [
                    [
                        {
                            "id": "test"
                        },
                        "test__volume.csv"
                    ]
                ],
                "wm_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_wm.nii.gz"
                    ]
                ],
                "wm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_wm.nii.gz"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-30T16:36:51.005323"
    },
    "anatomical_segmentation - fslfast": {
        "content": [
            {
                "csf_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_csf.nii.gz"
                    ]
                ],
                "csf_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_csf.nii.gz"
                    ]
                ],
                "gm_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_gm.nii.gz"
                    ]
                ],
                "gm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_gm.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,fcaa7c88e19009349d51422fab2077fc"
                ],
                "wm_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_wm.nii.gz"
                    ]
                ],
                "wm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_wm.nii.gz"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-30T16:36:14.558096"
    },
    "anatomical_segmentation - both": {
        "content": [
            {
                "csf_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_csf.nii.gz"
                    ]
                ],
                "csf_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_csf.nii.gz"
                    ]
                ],
                "gm_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_gm.nii.gz"
                    ]
                ],
                "gm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_gm.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,7f184d3df98a22ec61c2e52230199180",
                    "versions.yml:md5,fcaa7c88e19009349d51422fab2077fc"
                ],
                "wm_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_wm.nii.gz"
                    ]
                ],
                "wm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_wm.nii.gz"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-30T16:36:44.805357"
    }
}